### PR TITLE
Handle empty bodies safely

### LINKED ISSUE #2350: Handle empty function bodies safely
Currently an empty function body is accepted if a function returns a non-`None` value (even with `--strict-optional`). This is mostly useful for abstract methods. This is unsound, since we don't prevent anybody from calling these empty functions. 

Here are some ideas for making this safe:
- Only allow empty function bodies for abstract methods. Other functions will need something like a return statement, a raise statement or `assert False` or `# type: ignore`.
- Tag abstract methods with an empty body with a flag in the AST, and disallow calls to them via `super()`.

This still wouldn't address empty functions defined in stubs, but perhaps can just assume that every function in a stub has a non-empty body.

See also discussion in #2339.

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes #2350

This essentially re-applies https://github.com/python/mypy/pull/8111 modulo various (logical) changes in master since then.The only important difference is that now I override few return-related error codes for empty bodies, to allow opting out easily in next few versions (I still keep the flag to simplify testing).

### CODE REVIEW

**Review by hauntsaninja — COMMENTED** [review]
primer fallout seems really bad :-/ there's at least one fixable issue related to attrs, e.g. https://github.com/sco1/pylox/blob/393c029a8f503459e36367b990e43736f04bc5b9/pylox/tokens.py#L66

**Comment by JukkaL** [comment]
I think that it will be inevitable that this will generate some new errors, and it seems worth it since not having the checks has been a pretty big omission in mypy (of course assuming that they aren't false positives).

File: mypy/checker.py [hunk 1]
```diff
@@ -1202,28 +1210,81 @@ def check_func_def(self, defn: FuncItem, typ: CallableType, name: str | None) ->
...
 
+                allow_empty = allow_empty or self.options.allow_empty_bodies
+
+                show_error = (
+                    not body_is_trivial
+                    or
+                    # Allow empty bodies for abstract methods, overloads, in tests and stubs.
+                    not allow_empty
+                    and not (isinstance(defn, FuncDef) and defn.abstract_status != NOT_ABSTRACT)
+                    and not self.is_stub
```
Comment (JukkaL) [inline mypy/checker.py:1221]: Can you parens around the top-level expressions to make precedence clearer?

File: mypy/checker.py [hunk 2]
```diff
@@ -1202,28 +1210,81 @@ def check_func_def(self, defn: FuncItem, typ: CallableType, name: str | None) ->
...
+                # We want to minimize the fallout from checking empty bodies
+                # that was absent in many mypy versions.
+                if body_is_trivial and is_subtype(NoneType(), return_type):
+                    show_error = False
+
+                may_be_abstract = (
+                    body_is_trivial
+                    and defn.info is not FUNC_NO_INFO
+                    and defn.info.metaclass_type is not None
+                    and defn.info.metaclass_type.type.fullname == "abc.ABCMeta"
```
Comment (JukkaL) [inline mypy/checker.py:1244]: Should we use something like `has_base("abc.ABCMeta")` instead?

File: mypy/checker.py [hunk 3]
```diff
@@ -1202,28 +1210,81 @@ def check_func_def(self, defn: FuncItem, typ: CallableType, name: str | None) ->
...
-                    if not self.current_node_deferred and not isinstance(
-                        return_type, (NoneType, AnyType)
+                    if (
+                        not self.current_node_deferred
+                        and not isinstance(return_type, (NoneType, AnyType))
+                        and show_error
                     ):
                         # Control flow fell off the end of a function that was
                         # declared to return a non-None type and is not
                         # entirely pass/Ellipsis/raise NotImplementedError.
```
Comment (JukkaL) [inline mypy/checker.py:1255]: This comment may need to be updated.

File: mypy/checker.py [hunk 4]
```diff
@@ -1202,28 +1210,81 @@ def check_func_def(self, defn: FuncItem, typ: CallableType, name: str | None) ->
...
                         else:
-                            self.fail(message_registry.MISSING_RETURN_STATEMENT, defn)
-                else:
+                            msg = message_registry.MISSING_RETURN_STATEMENT
+                        if body_is_trivial:
+                            msg = msg._replace(code=codes.EMPTY_BODY)
+                        self.fail(msg, defn)
+                        if may_be_abstract:
+                            self.note(
+                                "If the method is meant to be abstract use @abc.abstractmethod",
```
Comment (JukkaL) [inline mypy/checker.py:1266]: Nit: Add a comma: "... be abstract, use ..."

File: mypy/checker.py [hunk 5]
```diff
@@ -1202,28 +1210,81 @@ def check_func_def(self, defn: FuncItem, typ: CallableType, name: str | None) ->
...
+                            subtype=NoneType(),
+                            supertype_label="expected",
+                            supertype=return_type,
+                            context=defn,
+                            msg=msg,
+                        )
+                        and may_be_abstract
+                    ):
+                        self.note(
+                            "If the method is meant to be abstract use @abc.abstractmethod", defn
```
Comment (JukkaL) [inline mypy/checker.py:1286]: Similar to above, and also can you use the message by putting it in a constant in `mypy.error_registry`?

File: mypy/main.py [hunk 6]
```diff
@@ -999,6 +999,9 @@ def add_invertible_flag(
         "the contents of SHADOW_FILE instead.",
     )
     add_invertible_flag("--fast-exit", default=True, help=argparse.SUPPRESS, group=internals_group)
+    add_invertible_flag(
```
Comment (JukkaL) [inline mypy/main.py:1002]: Add comment explaining why we have this.

File: mypy/nodes.py [hunk 7]
```diff
@@ -781,6 +781,8 @@ class FuncDef(FuncItem, SymbolNode, Statement):
         "abstract_status",
         "original_def",
         "deco_line",
+        "is_trivial_body",
+        "is_mypy_only",
```
Comment (JukkaL) [inline mypy/nodes.py:785]: Should this be added to `FUNCDEF_FLAGS`?

File: test-data/unit/fine-grained.test [hunk 8]
```diff
@@ -10031,3 +10031,46 @@ class C(B): ...
...
+from typing import Protocol
+class Base(Protocol):
+    def meth(self) -> int: return 0
+[file b.py.2]
+from typing import Protocol
+class Base(Protocol):
+    def meth(self) -> int: ...
+[out]
+==
+main:6: error: Call to abstract method "meth" of "Base" with trivial body via super() is unsafe
```
Comment (JukkaL) [inline test-data/unit/fine-grained.test:10076]: Maybe add incremental mode tests to check that serialization of the new flag (or flags) works.

**Review by JukkaL — APPROVED** [review]
Looks great, just left some minor comments. This fixes a big safety hole in type checking. Thanks for carefully minimizing the fallout and false positives.

File: mypy/checker.py (hunk 2 above)
Comment (ilevkivskyi) [reply mypy/checker.py:1244]: Yes, good point!

File: mypy/nodes.py (hunk 7 above)
Comment (ilevkivskyi) [reply mypy/nodes.py:785]: Yes, good catch, we should serialize this

**Pull request merged**